intrinsics/fnsub.go

```diff
@@ -1,21 +1,27 @@
 package intrinsics
 
 import (
+	"regexp"
 	"strings"
 )
 
 // FnSub resolves the 'Fn::Sub' AWS CloudFormation intrinsic function.
 // See: http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference-sub.html
 func FnSub(name string, input interface{}, template interface{}) interface{} {
 
+	// Input can either be a string for this type of Fn::Sub call:
+	// { "Fn::Sub": "some-string-with-a-${variable}" }
+
+	// or it will be an array of length two for named replacements
 	// { "Fn::Sub": [ "some ${replaced}", { "replaced": "value" } ] }
 
-	// Check the input is an array
-	if arr, ok := input.([]interface{}); ok {
-		// The first element is the source
-		if src, ok := arr[0].(string); ok {
+	switch val := input.(type) {
+
+	case []interface{}:
+		// Replace each of the variables in element 0 with the items in element 1
+		if src, ok := val[0].(string); ok {
 			// The seconds element is a map of variables to replace
-			if replacements, ok := arr[1].(map[string]interface{}); ok {
+			if replacements, ok := val[1].(map[string]interface{}); ok {
 				// Loop through the replacements
 				for key, replacement := range replacements {
 					// Check the replacement is a string
```

Comment:
I was saying in the CloudFormation intrinsic language - !Ref and !GetAtt return values of runtime parameters. All other intrinsics operate on the values returned by them - like !Sub, !Join etc. so as long as we document !Ref and !GetAtt behavior, users can easily infer the behavior of other intrinsics